### ISSUE TITLE
Races on the FileData struct?

### ISSUE THREAD

**Issue by corentone** [spf13/afero#133]
Hello, 

I'm getting a raceCondition between Read and Write to a file on the MemFS, see the snippet below:
```
WARNING: DATA RACE
Read at 0x00c42009a318 by goroutine 7:
 SOMEPATH/vendor/github.com/spf13/afero/mem.(*FileInfo).Size()
      SOMEPATH/vendor/github.com/spf13/afero/mem/file.go:275 +0x9d
  SOMEPATH/vendor/github.com/spf13/afero.ReadFile()
      SOMEPATH/vendor/github.com/spf13/afero/ioutil.go:76 +0x1bf
  SOMEPATH.readFiles()

Previous write at 0x00c42009a318 by goroutine 6:
  SOMEPATH/vendor/github.com/spf13/afero/mem.(*File).Write()
      SOMEPATH/vendor/github.com/spf13/afero/mem/file.go:236 +0x98c
  SOMEPATH/vendor/github.com/spf13/afero.WriteFile()
     SOMEPATH/vendor/github.com/spf13/afero/ioutil.go:129 +0xcb
  SOMEPATH.writeFile()
```

(I do have a routine reading the file and one writing it).

Looking deeper at the code, I'm slightly confused by the locking code in https://github.com/spf13/afero/blob/master/mem/file.go

Reading the name on *FileData is being locked (https://github.com/spf13/afero/blob/master/mem/file.go#L62) but writing it is not (https://github.com/spf13/afero/blob/master/mem/file.go#L76).

I feel a whole lots of locking are missing here or I could be missing the point of the lock?!

**Comment by jszwec** [comment]
I believe that my pull request fixes this. Unfortunately It has been opened for a while now.
https://github.com/spf13/afero/pull/128

**Comment by corentone** [comment]
@jszwec we might actually need both #128 and #134? If one is accessing any FileInfo directly to change the file then it would also trigger. Now we would need to make sure we don't double lock.

I like that you've added a test too. I should do that too. Let me try to ping a recent contributor if we can move forward on that.

**Comment by corentone** [comment]
@bep @mbertschler (you are the two most recent merging members for PRs so pinging you), could you weight on this? It would be awesome to get those races fixed.

**Comment by mbertschler** [comment]
Fixed by #134
